core/editor.js

```diff
@@ -168,8 +168,23 @@
 		} );
 		this.on( 'mode', updateCommands );
 
+		// (#3175)
+		this.on( 'selectionCheck', function( evt ) {
+			if ( evt.data ) {
+				evt.data.optimizeInElementEnds();
+			}
+		} );
+
 		// Handle startup focus.
 		this.on( 'instanceReady', function() {
+			this.editable().on( 'keydown', function( evt ) {
```

Comment:
My concluding thought about not optimizing when <kbd>Shift</kbd> is pressed was that it should be combined with <kbd>Arrow key</kbd> presses so we should prevent optimization only when <kbd>Shift</kbd> + <kbd>Arrow key</kbd> was pressed - I see it never appeared in my previous https://github.com/ckeditor/ckeditor-dev/pull/3213#pullrequestreview-276645022 comment 🤔 

So this way all other combinations with <kbd>Shift</kbd> will trigger optimization, but it will still solve the issue with <kbd>Arrow key</kbd>s, WDYT?

It's not a big change, since instead of `keydown` you will need `key` event checking `keyCode` - the same way it was done in https://github.com/ckeditor/ckeditor-dev/pull/3213/commits/b8901c2eaf138e87e4a7f7d6db500862731dff46#diff-162a1b36d2efd62081522822a026e853R173. And resetting `shiftPressed` flag on `keyup` will still work.

Reply:
After discussion we leave it as it is.